migration/identity-migration/src/main/java/io/camunda/migration/identity/TenantMappingRuleMigrationHandler.java

```diff
@@ -0,0 +1,84 @@
+/*
+ * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH under
+ * one or more contributor license agreements. See the NOTICE file distributed
+ * with this work for additional information regarding copyright ownership.
+ * Licensed under the Camunda License 1.0. You may not use this file
+ * except in compliance with the Camunda License 1.0.
+ */
+package io.camunda.migration.identity;
+
+import io.camunda.migration.identity.dto.Tenant;
+import io.camunda.migration.identity.dto.TenantMappingRule;
+import io.camunda.migration.identity.midentity.ManagementIdentityClient;
+import io.camunda.migration.identity.midentity.ManagementIdentityTransformer;
+import io.camunda.migration.identity.midentity.MigrationStatusUpdateRequest;
+import io.camunda.migration.identity.service.MappingService;
+import io.camunda.migration.identity.service.TenantService;
+import io.camunda.search.entities.TenantEntity;
+import java.util.List;
+import org.springframework.stereotype.Component;
+
+@Component
+public class TenantMappingRuleMigrationHandler implements MigrationHandler {
+
+  public static final int SIZE = 100;
+  private final ManagementIdentityClient managementIdentityClient;
+  private final ManagementIdentityTransformer managementIdentityTransformer;
+  private final TenantService tenantService;
+  private final MappingService mappingService;
+
+  public TenantMappingRuleMigrationHandler(
+      final ManagementIdentityClient managementIdentityClient,
+      final ManagementIdentityTransformer managementIdentityTransformer,
+      final TenantService tenantService,
+      final MappingService mappingService) {
+    this.managementIdentityClient = managementIdentityClient;
+    this.managementIdentityTransformer = managementIdentityTransformer;
+    this.tenantService = tenantService;
+    this.mappingService = mappingService;
+  }
+
+  @Override
+  public void migrate() {
+    TenantMappingRule lastTenantMappingRule = null;
+    do {
+      final List<TenantMappingRule> tenantMappingRules =
+          managementIdentityClient.fetchTenantMappingRules(lastTenantMappingRule, SIZE);
+      managementIdentityClient.updateMigrationStatus(
+          tenantMappingRules.stream().map(this::createTenantMappingRule).toList());
+      lastTenantMappingRule =
+          tenantMappingRules.isEmpty()
+              ? null
+              : tenantMappingRules.get(tenantMappingRules.size() - 1);
+    } while (lastTenantMappingRule != null);
+  }
+
+  private MigrationStatusUpdateRequest createTenantMappingRule(
+      final TenantMappingRule tenantMappingRule) {
+    try {
+      final var mappingKey =
+          mappingService.findOrCreateMapping(
+              tenantMappingRule.getName(),
+              tenantMappingRule.getClaimName(),
+              tenantMappingRule.getClaimValue());
```

Comment:
because here we need the mapping key for next step (assign mapping to tenant) to be available, so here conflict is a error and we need to retry. ( I have to add that in this case it's very unlikely to have a conflict because a search query is executed before and it's just for very rare case that the mapping is created but is not  published to elastic yet.)

Reply:
But the retry would only occur because the migration app crashes, right? I think we should try to handle this explicitly and not rely on the crash.

Reply:
Actually not, retry will happens automatically on next loop during fetch mapping rules. The endpoints in identity designed to return failed or not already migrated errors. To be precise it was designed like this and I applied a fix in next PR to actually achieve it.